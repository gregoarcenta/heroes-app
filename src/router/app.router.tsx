import { createBrowserRouter, Navigate } from "react-router";
import { SearchPage } from "@/heroes/pages/search/SearchPage";
import { HomePage } from "@/heroes/pages/home/HomePage";
import { HeroPage } from "@/heroes/pages/hero/HeroPage";
import { AdminPage } from "@/admin/pages/AdminPage";

export const appRouter = createBrowserRouter([
  {
    path: "/",
    element: <HomePage />
  },
  {
    path: "/hero/1",
    element: <HeroPage />
  },
  {
    path: "/search",
    element: <SearchPage />
  },
  {
    path: "admin",
    element: <AdminPage />
  },
  {
    path: "*",
    element: <Navigate to={"/"} />
  }
]);
