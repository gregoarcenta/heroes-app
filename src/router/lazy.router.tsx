import React from "react";

export const SearchPage = React.lazy(
  () => import("@/heroes/pages/search/SearchPage")
);

export const HeroPage = React.lazy(
  () => import("@/heroes/pages/hero/HeroPage")
);
