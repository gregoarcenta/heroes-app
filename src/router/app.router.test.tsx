import { describe, expect, test, vi } from "vitest";
import { appRouter } from "./app.router";
import {
  createMemoryRouter,
  Outlet,
  RouterProvider,
  useParams,
} from "react-router";
import { render, screen } from "@testing-library/react";

vi.mock("@/heroes/layouts/HeroesLayout", () => ({
  HeroesLayout: () => (
    <div data-testid="heroes-layout">
      <Outlet />
    </div>
  ),
}));

vi.mock("@/heroes/pages/home/HomePage", () => ({
  default: () => <div data-testid="home-page"></div>,
}));

vi.mock("@/heroes/pages/hero/HeroPage", () => ({
  default: () => {
    // eslint-disable-next-line react-hooks/rules-of-hooks
    const { slugId = "" } = useParams();
    return <div data-testid="hero-page">HeroPage - {slugId}</div>;
  },
}));

vi.mock("@/heroes/pages/search/SearchPage", () => ({
  default: () => <div data-testid="search-page"></div>,
}));

describe("appRouter", () => {
  test("should be configured as expected", () => {
    expect(appRouter.routes).toMatchSnapshot();
  });

  test("should render home page at root path", () => {
    const router = createMemoryRouter(appRouter.routes, {
      initialEntries: ["/"],
    });
    render(<RouterProvider router={router} />);
    expect(screen.getByTestId("home-page")).toBeDefined();
  });

  test("should render hero page at /heroes/:idSlug path", async () => {
    const router = createMemoryRouter(appRouter.routes, {
      initialEntries: ["/hero/superman"],
    });
    render(<RouterProvider router={router} />);
    expect(
      await screen.findByTestId("hero-page").then((r) => r.innerHTML),
    ).toContain("superman");
  });

  test("should render search page at /search path", async () => {
    const router = createMemoryRouter(appRouter.routes, {
      initialEntries: ["/search"],
    });

    render(<RouterProvider router={router} />);

    expect(await screen.findByTestId("search-page")).toBeDefined();
  });

  test("should redirect to home page for unknown routes", () => {
    const router = createMemoryRouter(appRouter.routes, {
      initialEntries: ["/otra-pagina-rara"],
    });

    render(<RouterProvider router={router} />);

    expect(screen.getByTestId("home-page")).toBeDefined();
  });
});
