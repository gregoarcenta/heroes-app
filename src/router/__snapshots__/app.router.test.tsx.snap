// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`appRouter > should be configured as expected 1`] = `
[
  {
    "children": [
      {
        "element": <default />,
        "hasErrorBoundary": false,
        "id": "0-0",
        "index": true,
      },
      {
        "children": undefined,
        "element": <UNDEFINED />,
        "hasErrorBoundary": false,
        "id": "0-1",
        "path": "hero/:slugId",
      },
      {
        "children": undefined,
        "element": <UNDEFINED />,
        "hasErrorBoundary": false,
        "id": "0-2",
        "path": "search",
      },
      {
        "children": undefined,
        "element": <Navigate
          to="/"
        />,
        "hasErrorBoundary": false,
        "id": "0-3",
        "path": "*",
      },
    ],
    "element": <HeroesLayout />,
    "hasErrorBoundary": false,
    "id": "0",
    "path": "/",
  },
  {
    "children": [
      {
        "element": <AdminPage />,
        "hasErrorBoundary": false,
        "id": "1-0",
        "index": true,
      },
    ],
    "element": <AdminLayout />,
    "hasErrorBoundary": false,
    "id": "1",
    "path": "admin",
  },
]
`;
